components/terrain/compositemaprenderer.cpp

```diff
@@ -59,26 +59,29 @@ void CompositeMapRenderer::drawImplementation(osg::RenderInfo &renderInfo) const
 
     while (!mImmediateCompileSet.empty())
     {
-        CompositeMap* node = *mImmediateCompileSet.begin();
+        osg::ref_ptr<CompositeMap> node = *mImmediateCompileSet.begin();
         mCompiled.insert(node);
+        mImmediateCompileSet.erase(node);
 
+        mMutex.unlock();
         compile(*node, renderInfo, nullptr);
-
-        mImmediateCompileSet.erase(mImmediateCompileSet.begin());
+        mMutex.lock();
     }
 
     double timeLeft = availableTime;
 
     while (!mCompileSet.empty() && timeLeft > 0)
     {
-        CompositeMap* node = *mCompileSet.begin();
+        osg::ref_ptr<CompositeMap> node = *mCompileSet.begin();
 
+        mMutex.unlock();
         compile(*node, renderInfo, &timeLeft);
+        mMutex.lock();
 
         if (node->mCompiled >= node->mDrawables.size())
         {
             mCompiled.insert(node);
-            mCompileSet.erase(mCompileSet.begin());
+            mCompileSet.erase(node);
```

Comment:
If we have only one `osg::GraphicsThread` to call this method, then erase here is fine. But if we have more, then two threads can start compiling same composite map. I'm not sure do they need to do it for each `osg::GraphicsContext` or not.

Reply:
@AnyOldName3, do we have several graphics threads?

Reply:
As a possible solution, we can try to remove composite map from the `mCompileSet` immediately as a draw thread starts to compile it. If we did not compile it fully (because of not enough time), place it back after processing. In this case graphics threads should not compile maps which we already compile in other threads.

Reply:
We use the DrawThreadPerContext threading model, so as the engine only ever has one context active at a time, we've only got one graphics thread (assuming that's how it works).